q2_sample_classifier/classify.py

```diff
@@ -38,6 +41,47 @@
 }
 
 
+def classify_samples_from_dist(ctx, dmtx, metadata, k,
+                               palette=defaults['palette']):
+    ''' Returns knn classifier results from a distance matrix.'''
+    distance_matrix = dmtx.view(skbio.DistanceMatrix)
+    predictions = []
+    metadata_series = metadata.to_series()
+    for i, row in enumerate(distance_matrix):
+        dists = []
+        categories = []
+        for j, dist in enumerate(row):
+            if j == i:
+                continue  # exclude self
+            dists.append(dist)
+            categories.append(metadata_series[distance_matrix.ids[j]])
+
+        # k-long series of (category: dist) ordered small -> large
```

Comment:
oh cool — looks like you just built your own knn classifier from scratch. Any reason not to use sklearn?

Reply:
yeah... Don't love the reinvention of knn here. But seems that sklearn's knn doesn't accept a distance matrix [1], and I want this to be usable with e.g. unifrac.

Could reach into the more private parts of sklearn though, or better would probably be submit a PR to sklearn with knn_from_distance_matrix

[1] http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier

Reply:
if metric == precomputed, it takes a distance matrix according to that link

But I'm not sure exactly how this would work for prediction.

It's a fairly simple method, so I don't think there's a problem with the reinvention here... this can be a baseline that we can use to implement sklearn's knn classifier later on if we see some advantage.

Reply:
Aye, it may be just my unfamiliarity with using sklearn's knn.